### PR TITLE
@clerk/upgrade detect binaries and gitignore during scan

### LINKED ISSUE #8323: @clerk/upgrade stalls on binary build artifacts
### Preliminary Checks

- [x] I have reviewed the documentation: https://clerk.com/docs

- [x] I have searched for existing issues: https://github.com/clerk/javascript/issues

- [x] I have not already reached out to Clerk support via email or Discord (if you have, no need to open an issue here)

- [x] This issue is not a question, general help request, or anything other than a bug report directly related to Clerk. Please ask questions in our Discord community: https://clerk.com/discord.

### Reproduction

https://gist.github.com/rofinn/34747e70ad8ecadd03f76d0d4200dfd9

### Publishable key

N/A

### Description

Running `clerk/upgrade --dry-run` in a monorepo containing cargo build artifacts causes the scan step to stall trying to analyze binary build files. The linked github gist includes the steps to reproduce with shell output.

**Steps to reproduce:**

1. Clone https://github.com/Nipsuli/tauri-plugin-clerk.git (or any repo that may produced binary build artifacts)
2. Generated the build artifacts. In the example, this was running `cargo build` in `examples/react-example/src-tauri`.
3. Run `time bun x @clerk/upgrade --dry-run` from the `examples/react-example` directory (to mimic a monorepo).
4. Kill the scan after it stalls
5. Delete the `examples/react-example/src-tauri/target/debug` directory and rerun
6. You'll notice it completes fine.

**Expected behavior:**

I would have expected the scanning step to focus on text files that contain potential code issues. I understand that just relying on extensions isn't the most reliable if you want to catch issues in scripts or markdown files, but I think it should be able to identify and catch binary build artifacts.

**Actual behavior:**

I think the desired behaviour has 4 components.

1. The scan file spinner should probably give the full relative path to make it easier to spot when this is happening. Some build artifacts can have pretty opaque names.
2. I think we should be able to detect binary files as part of the scan.
3. (Optional) Broaden the default ignore globs include include common build output directories
4. (Optional) It'd be nice if the scanner picked up existing .gitignore rules

### Environment

```Shell
N/A. Can be reproduced in any repo with binary build artifacts.
```

### PR DESCRIPTION
**Pull request by rofinn**
## Description

Fixes #8323 (sorry didn't notice you prefer having issues created first)

As outlined in the issue and linked gist to reproduce the issue, this PR attempts to address 3 particular issues faced while using the `@clerk/upgrade` cli.

1. The scanning filename spinner should give the full relative path to make debugging easier.
2. Uses [isbinaryfile](https://github.com/gjtorikian/isBinaryFile) to detect if a candidate file is binary (primary issue from #8323 )
3. (Optional) Adds support for using `.gitignore` files to reduce the scanning scope. And falls through if `git` is not available or no `.gitignore` file is present.

NOTES:
- Both (2) and (3) include additional tests for detecting binary files, .gitignore files, along with the fallback cases like when `git` is not available (or not a repo) or if `--skip-ignore` is used.
- I opted to add isbinaryfile as a dependency to packages/upgrade` since it covers more edge cases than my initial implementation.
- As per the [tinyglobby docs](https://superchupu.dev/tinyglobby/migration), I'm just shelling-out for the gitignore check.

## Checklist

- [X] `pnpm test` runs as expected.
- [X] `pnpm build` runs as expected.
- [ ] (If applicable) [JSDoc comments](https://jsdoc.app/about-getting-started.html) have been added or updated for any package exports
- [ ] (If applicable) [Documentation](https://github.com/clerk/clerk-docs) has been updated

Just updated the README and the CLI help output.

## Type of change

- [ ] 🐛 Bug fix
- [ ] 🌟 New feature
- [ ] 🔨 Breaking change
- [ ] 📖 Refactoring / dependency upgrade / documentation
- [X] other: I'd mostly call this a bug fix, but it does add an additional dependency, and the gitignore behaviour could be a breaking change.

### CODE REVIEW

**Comment by rofinn** [comment]
NOTE: I have not created a changeset yet because I think the version bump may depend on whether the last commit (gitignore) is included.

File: packages/upgrade/src/__tests__/integration/runner.test.js [hunk 1]
```diff
@@ -219,6 +233,132 @@ describe('runScans', () => {
...
+      ignore: [],
+    });
+
+    const expected = path.relative(process.cwd(), path.join(fixture.path, 'run'));
+    expect(results[0].instances.some(instance => instance.file === expected)).toBe(true);
+  });
+
+  it('skips binary files that are not covered by ignore globs', async () => {
+    const config = await loadConfig('nextjs', 6);
+    config.changes = config.changes.filter(change => change.matcher);
```
Comment (rofinn) [inline packages/upgrade/src/__tests__/integration/runner.test.js]: I got a bit lazy while testing. If folks want I can add a concrete config for each test to be more specific?

File: packages/upgrade/src/__tests__/integration/runner.test.js [hunk 2]
```diff
@@ -219,6 +233,132 @@ describe('runScans', () => {
...
+
+    const binaryPath = path.join(fixture.path, 'binary');
+    fs.writeFileSync(binaryPath, Buffer.from([0x7f, 0x45, 0x4c, 0x46, 0x00]));
```
Comment (rofinn) [inline packages/upgrade/src/__tests__/integration/runner.test.js]: I guess this would be a stronger test if I added `Buffer.from('afterSignInUrl')` to it.

**Review by rofinn — COMMENTED** [review]
Just noticed a few test cleanup items on a second review.

File: packages/upgrade/src/__tests__/integration/runner.test.js [hunk 3]
```diff
@@ -219,6 +247,107 @@ describe('runScans', () => {
...
+    fs.writeFileSync(path.join(fixture.path, 'run'), '#!/usr/bin/env node\nafterSignInUrl\n', 'utf8');
+
+    const results = await runScans(createScanConfig(), 'nextjs', {
+      dir: fixture.path,
+      ignore: [],
+    });
+
+    const file = path.relative(process.cwd(), path.join(fixture.path, 'run'));
+    expect(results.flatMap(result => result.instances).some(instance => instance.file === file)).toBe(true);
+  });
```
[the remark this answers is not in the document]
Comment (rofinn) [reply packages/upgrade/src/__tests__/integration/runner.test.js:369]: Latest commit checks the `renderDebug` output when `git ls-files` exceeds the max child process buffer. Didn't do this for all existing tests though.

File: packages/upgrade/src/runner.js [hunk 4]
```diff
@@ -156,6 +158,38 @@ function loadMatchers(config, sdk) {
...
+  function gitIgnored() {
+    try {
+      return execSync('git ls-files --others --ignored --exclude-standard --directory', {
+        cwd,
+        encoding: 'utf8',
+        stdio: ['ignore', 'pipe', 'ignore'],
+      })
+        .split('\n')
+        .filter(Boolean)
+        .map(file => escapePath(file));
```
[the remark this answers is not in the document]
Comment (rofinn) [reply packages/upgrade/src/runner.js]: Sure. Changed with a test case for "unusual" characters per the manpage. I'm not sure if either of these additional review comments/commits is worth it, but I guess we're appeasing the AI overlords now 🤷🏻

**Comment by rofinn** [comment]
Closing, as I'm moving off clerk and don't feel inclined to continue updating this PR. Feel free to reopen or copy this if someone thinks this addition would be useful.
